diesel/src/sqlite/connection/raw.rs

```diff
@@ -194,12 +251,150 @@ extern "C" fn run_custom_function<F>(
+            // only access the tag of an zeroed variant here, right? That tag is guaranteed to be zero for
+            // our case and this variant.
+            Some(ref mut a) => match a.as_mut() {
+                &mut OptionalAggregator::Some(ref mut agg) => agg,
+                &mut OptionalAggregator::None => {
+                    // Not sure if we should use `write_unaligned` instead
+                    ptr::write(a.as_mut(), OptionalAggregator::Some(A::default()));
+                    if let &mut OptionalAggregator::Some(ref mut agg) = a.as_mut() {
+                        agg
+                    } else {
+                        unreachable!("We've written the aggregator above to that location, it must be there")
+                    }
+                }
+            },
+            None => {
+                ffi::sqlite3_result_error(
+                    ctx,
+                    NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
+                    NULL_AG_CTX_ERR.len() as _,
+                );
+                return;
+            }
+        };
+
+        let mut f = |args: &[*mut ffi::sqlite3_value]| {
+            let mut row = FunctionRow { args };
+            let args_row = Args::Row::build_from_row(&mut row).map_err(Error::DeserializationError).unwrap(); // TODO @thekuom: error handling
+            let args = Args::build(args_row);
+
+            aggregator.step(args);
+        };
+
+        let args = slice::from_raw_parts(value_ptr, num_args as _);
+        f(args);
+    }
+}
+
+extern "C" fn run_aggregator_final_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
+    ctx: *mut ffi::sqlite3_context,
+)
+where
+    A: Aggregator<Args, Output=Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    static NULL_AG_CTX_ERR: &str = "An unknown error occurred. sqlite3_aggregate_context returned a null pointer. This should never happen.";
+
+    unsafe {
+        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, 0);
+        let aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        // TODO @thekuom: abstract this out
+        let aggregator = match aggregate_context {
+            // Check with someone that does more unsafe code reviews if this is ok. We do
+            // only access the tag of an zeroed variant here, right? That tag is guaranteed to be zero for
+            // our case and this variant.
+            Some(ref a) => match a.as_ref() {
+                &OptionalAggregator::Some(ref agg) => agg,
+                &OptionalAggregator::None => {
+                    ffi::sqlite3_result_error(
+                        ctx,
+                        NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
+                        NULL_AG_CTX_ERR.len() as _,
+                        );
+                    return;
+                }
+            },
```

Comment:
Forget what I've wrote above, that solution will leak memory.
We need to move our aggregator context out of the the sqlite context field and explicitly call drop on it. That means the code above should look like this:
```suggestion
            Some(ref a) => match std::mem::replace(a.as_mut(), OptionalAggregator::None {
                OptionalAggregator::Some(agg) => agg,
                OptionalAggregator::None => {
                    ffi::sqlite3_result_error(
                        ctx,
                        NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
                        NULL_AG_CTX_ERR.len() as _,
                        );
                    return;
                }
            },
```

so that in the end of the function our context is cleared (== set to `None`) and the destructor of our aggregator, so that potential allocated memory/… is freed.